lib/cli/cli_scrn.c

```diff
@@ -232,4 +228,8 @@ RTE_INIT(scrn_constructor)
 	memset(scrn, 0, sizeof(struct cli_scrn));
 
 	this_scrn = scrn;
+
+	scrn_set_io(stdin, stdout);
+	memset(&this_scrn->oldterm, 0, sizeof(struct termios));
```

Comment:
After looking at the code a bit, you are correct it was wrong, thanks.

The question I have is do I need to do the second tcgetattr() call, but instead remove the memset/tcgetattr() calls and replace with:
memcpy(&this_scrn->oldterm, &term, sizeof(struct termios));

If this is acceptable I can make that change or you can update the PR, let me know.

Thanks again.

Reply:
The second tcgetattr is used for adjusting the c_lflags:
https://github.com/pktgen/Pktgen-DPDK/blob/20f0571bf3eca850a16c5a856b8fde1d27532059/lib/cli/cli_scrn.c#L106

If you just copy the term to oldterm, the original terminal flags will lost.  According to my understanding, the oldterm is used for storeing the original flags before launching pktgen and will be restored when quit. 
Any misunderstanding?

Reply:
It looks like my message dropped some text or something. Here it is again.

I had a bunch more text here, but lost it. The goal is to add the code to get the termios in the scrn_constructor() routine and use oldterm to setup stdin later.  Have a read and let me know if that works for you.

``` c
diff --git lib/cli/cli_scrn.c lib/cli/cli_scrn.c
index 88417d4..1daa084 100644
--- lib/cli/cli_scrn.c
+++ lib/cli/cli_scrn.c
@@ -100,14 +100,7 @@ scrn_stdin_setup(void)
        if (!scrn)
                return -1;

-       scrn_set_io(stdin, stdout);
-
-       memset(&scrn->oldterm, 0, sizeof(term));
-       if (tcgetattr(fileno(scrn->fd_in), &scrn->oldterm) ||
-           tcgetattr(fileno(scrn->fd_in), &term)) {
-               fprintf(stderr, "%s: setup failed for tty\n", __func__);
-               return -1;
-       }
+       memcpy(&term, &scrn->oldterm, sizeof(term));

        term.c_lflag &= ~(ICANON | ECHO | ISIG | IEXTEN);

@@ -148,7 +141,6 @@ handle_winch(int sig)
        this_scrn->ncols = w.ws_col;

        /* Need to refresh the screen */
-       //cli_clear_screen();
        cli_clear_line(-1);
        cli_redisplay_line();
 }
@@ -185,8 +177,6 @@ scrn_create(int scrn_type, int theme)
        } else if (scrn_type == SCRN_NOTTY_TYPE) {
                scrn->nrows = 24;
                scrn->ncols = 80;
-
-               scrn_set_io(stdin, stdout);
        } else {
                fprintf(stderr, "%s: unexpected scrn_type %d\n",
                        __func__, scrn_type);
@@ -229,7 +219,11 @@ RTE_INIT(scrn_constructor)
                exit(-1);
        }

-       memset(scrn, 0, sizeof(struct cli_scrn));
-
        this_scrn = scrn;
+
+       scrn_set_io(stdin, stdout);
+       if (tcgetattr(fileno(scrn->fd_in), &scrn->oldterm)) {
+           fprintf(stderr, "%s: tcgetattr failed\n", __func__);
+               exit(-1);
+       }
 }
```

Reply:
Yes, I think this works. And another suggestion is removing this line in scrn_stdin_restore:
```
if (system("stty sane"))
                fprintf(stderr, "%s: system command failed\n", __func__);
```
The reason is, we have already backup the term and restored it. The "stty sane" will change the behavior to some DEFAULT setting, which is different from my customized setting.

Reply:
I merged the changes to main and should cover this PR, so this one can be closed.

Thank you for finding and pointing out this issue.